Allow picking of a random subfaction.

phrohdoh opened the pull request:
~~Works in ra.~~
~~Crashes in cnc/d2k/ts but I cannot figure out why.~~

~~Help much appreciated!~~

pchote commented:
Using magic race names for this feels rather hacky.  It might work out cleaner if you defined the concept of a random class, with each race opting in to one or more of them.

WolfGaming commented:
We could add a country group trait where you can add all the races you want in it.
For example
```
	CountryGroup:
		Name: Any bla
		Races: bla, bla2, bla3
```

we could even add it to the prereq system so any races in it get it as a provided prereq, but that is optional.

Phrohdoh commented:
That was actually my initial implement (the name was CountryGroup) but I couldn't get it working.

I'll try again tonight.

WolfGaming commented:
@Phrohdoh If you have any problems I can help, but I dont have time to code anything myself atm, maybe in a few weeks.

LipkeGu commented:
```
foreach (var c in selectableCountries)
			{
				var side = c.Side;
				if (side == null)
					continue;

				if (countries.Keys.Contains("random_{0}".F(side.ToLower())))
					continue;

				countries.Add(side, new LobbyCountry { Name = "Any", Side = c.Side });
			}
```

works for me... :)

WolfGaming commented:
@LipkeGu This however does not support having a race be a part of two groups. Also this would force mods and the like to add sides which makes very little sense if you only have 2 factions.

LipkeGu commented:
@WolfGaming : its the same logic as in the commit, i added only a check for null to prevent a crashes

LipkeGu commented:
i changed it again so it more "smarter"
```
foreach (var c in selectableCountries)
{
	var side = c.Side;
	if (side == null)
		continue;

	if (countries.Keys.Contains("random_{0}".F(side.ToLower())) || countries.Keys.Contains(c.Side))
		continue;
				
	countries.Add(side, new LobbyCountry { Name = "Any", Side = c.Side });
}
```

Phrohdoh commented:
I'm working on this again but would like input on the flag icons created here.
Yay/nay?

Unit158 commented:
You could also try just the try the any flag with the grey bits replaced by the side's respective colour (i.e., a question mark on a blue flag for Allies). Otherwise, it looks fine to me.

On 2015-01-28, at 11:03 PM, Taryn Hill <notifications@github.com> wrote:

> I'm working on this again but would like input on the flag icons created here.
> Yay/nay?
> 
> \
> Reply to this email directly or view it on GitHub.
>

Phrohdoh commented:
Meh, afterall I prefer the current flags personally.

**Update:**
The remaining steps are:
• rename oramap -> zip
• unzip
• `s/"Race: Random"/"Race: random"/`
• rezip
• rename zip -> oramap.

Phrohdoh commented:
Updated the maps.
Ready for review.

phrohdoh committed:
Add random_soviet and random_allies flags.
Add World trait CountryGroup.
Generate LobbyCountry for each CountryGroup.
Session.Country -> CountryRace for clarity.

RoosterDragon commented on OpenRA.Game/Player.cs:
```diff
@@ -46,14 +46,31 @@ public class Player : IScriptBindable, IScriptNotifyBind, ILuaTableBinding, ILua
+					throw new YamlException("No CountryGroup found with Race `{0}` on map `{1}`.".F(race, world.Map.Title));
+
+				var shuffled = group.RaceMembers.Shuffle(r).ToArray();
+				selected = selectableCountries.First(c => c.Race == shuffled[0]);
```
Looks like Random would suffice here (unless I'm missing something).
```
var randomRace = group.RaceMembers.Random(r);
selected = selectableCountries.First(c => c.Race == randomRace);
```

Phrohdoh replied:
You are right, thank you sir.

RoosterDragon commented:
This seems to be making the linter fall over (check the Travis logs).

pchote commented:
@Phrohdoh: sorry to break it to you, but updating the map yaml by hand isn't sufficient, as that doesn't account for the hundreds of maps on the resource site and any custom installed/created maps that players may have (as well as just being really inefficient).  The standard way to do this is to bump the yaml map format and add an on-load check/fix/save.

Phrohdoh commented:
@RoosterDragon that is due to the `Random` vs `random` yaml changes in maps.

Phrohdoh commented:
Updated (bumped MapFormat suggestion from @pchote).
Fixed a rare OutOfIndex introduced by older versions of this PR's code.

Mailaender commented on OpenRA.Game/Traits/World/CountryGroup.cs:
```diff
@@ -0,0 +1,28 @@
+//  * as published by the Free Software Foundation. For more information,
+//  * see COPYING.
+//  */
+// #endregion
```
The `// ` are a bit strange in front of multi-line comments. I know that MonoDevelop does them when you use a templated header which seems to be a bug.

Phrohdoh replied:
Fixed. Thanks I always mean to remove this before pushing but obviously forgot here.

abcdefg30 commented:
Needs a rebase.

phrohdoh committed:
Bump MapFormat to 8.

Phrohdoh commented:
Rebased and went with @Unit158's suggestion about the flags (should I change the outline color too?).
The flags aren't as pretty but they do get the point across better.

Fixed some typos (late night hacking).

Phrohdoh commented:
@Mailaender Could you please restart the AppVeyor build if possible?

Mailaender commented:
https://github.com/OpenRA/OpenRA/pull/7408 will make this less dependent on SourceForge mirror availability.

penev92 commented:
We really should start moving away from the Country/Race/CountryRaceFactionHouse names.
"Faction" is generic enough to combine all the others and we should start using that everywhere.
While I'm not saying it should be done in this PR, this is a nice place to start.

penev92 commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -102,7 +102,7 @@ public static Client Deserialize(MiniYaml data)
 			public HSLColor PreferredColor; // Color that the client normally uses from settings.yaml.
 			public HSLColor Color; // Actual color that the client is using. Usually the same as PreferredColor but can be different on maps with locked colors.
-			public string Country;
+			public string CountryRace;
```
`Country` to `Faction` please.
You are renaming this anyway, so might as well do it right the first time.

penev92 commented on OpenRA.Game/Traits/World/CountryGroup.cs:
```diff
@@ -0,0 +1,29 @@
+		public readonly string[] RaceMembers = { };
+	}
+
+	public class CountryGroup { }
```
`CountryGroup` -> `FactionGroup` please.

penev92 commented on OpenRA.Game/Traits/World/CountryGroup.cs:
```diff
@@ -0,0 +1,29 @@
+		public readonly string Race = null;
+
+		[Desc("Races that are a member of this group.")]
+		public readonly string[] RaceMembers = { };
```
"FactionMembers", perhaps?

penev92 commented on OpenRA.Game/Traits/World/CountryGroup.cs:
```diff
@@ -0,0 +1,29 @@
+
+		[Desc("Used for UI flags and other race-specific things.",
+		"Must not be shared with any other Countrys or CountryGroups.")]
+		public readonly string Race = null;
```
No idea about this one, but it shouldn't be `Race`.

Phrohdoh replied:
Sure it should be. It is used in the same way the other `Race`s are used (i.e. flags).

penev92 replied:
Nobody said they are correct.

Phrohdoh replied:
Okay, well they are that and I won't change this to match what those aren't.

pchote replied:
@penev92: you may not agree with it, but it is the convention used by OpenRA (and it's used in a lot of places).  This is not the place to try and get it changed.

pchote replied:
Actually... the conventional use of `Race` is actually what `RaceMembers` below is.  I agree that this probably shouldn't be called `Race`.  The description doesn't really clarify the matter, though.

penev92 commented on OpenRA.Mods.Common/Widgets/Logic/Lobby/LobbyLogic.cs:
```diff
@@ -142,9 +142,15 @@ void CloseWindow()
-				countries.Add(c.Race, new LobbyCountry { Name = c.Name, Side = c.Side, Description = c.Description });
+			var worldTraits = modRules.Actors["world"].Traits;
+
+			var countryGroupes = worldTraits.WithInterface<CountryGroupInfo>();
```
"GroupEs" ?
English, please :D

Phrohdoh replied:
derp

penev92 commented:
I believe that OpenRA has outgrown "Countries" years ago.
We have countries in RA, factions in TD/TS, Houses in D2k. Unofficial mods may include "civilizations" ( @Phrohdoh should know what I'm talking about) and "species".
The only term that is generic enough for all those is "Factions".
The translation to that should start somewhere, and I think naming the new additions you make with this PR in that style is a good start. Otherwise you are just creating more work for whoever decides to do it somewhere in the future.
Sure, it would be even less consistent for a while, with an additional term added to the mess, but the current state of things is inconsistent enough.

On a side note I reviewed it and it works. My only notes are about the naming. But from what I hear on IRC, the `Random` -> `random` renaming might break thirdparty maps?

pchote commented:
@penev92 the Race and Country terminology is used (consistently) through every part of the engine: from the low level map format and sound code through to the UI rendering and traits.  I disagree completely with the assertion that this is an appropriate place to start changing things.

pchote commented on OpenRA.Game/GameInformation.cs:
```diff
@@ -114,7 +114,7 @@ public void AddPlayer(OpenRA.Player runtimePlayer, Session lobbyInfo)
 				Team = client.Team,
 				SpawnPoint = runtimePlayer.SpawnPoint,
-				IsRandomFaction = runtimePlayer.Country.Race != client.Country,
+				IsRandomFaction = runtimePlayer.Country.Race != client.CountryRace,
```
What exactly is a CountryRace?

Phrohdoh replied:
It was used in the same manner as `Race` but I figured someone would get tripped up by that so I just appended `Race` to the field instead.

`Country.Race != Country` is a bit bogus IMO.

It's the best I could do in a bad situation.

pchote replied:
Oh, I see.  This is a historical naming fail: `client.Country` actually specifies a race (and probably wants to be renamed to `Race`).

Phrohdoh replied:
That I can do.

pchote commented:
This feature can be implemented with significantly less code changes:
* Add a `RandomRace` field to `CountryInfo`.
* Add three new `Country` definitions to the world actor:
```
 	Country@random:
 		Name: Any
 		Race: Random
 		Description: A random country.
 	Country@randomallies:
 		Name: Any Allied
 		Race: Random
 		Side: Allies
 		Description: A random Allied country.
 	Country@random:
 		Name: Any Soviet
 		Race: Random
  		Side: Soviet
		Description: A random Soviet country.
```
 * Change `Session` to define `Country = "random"` &rarr; `Race = "Random",`.

You then don't need any new traits or map format changes, which makes this significantly less controversial.

Phrohdoh commented:
Hmm, I'm really not a fan of that approach to this scenario. I feel it is more error prone on the yaml side.

I also don't like `Random` for the race because all other races are lowercase, so I'd make the map changes either way, personally. Up to you to close this now or wait for me to implement and like the above suggestion.

Or someone else can do it.

pchote commented:
> I feel it is more error prone on the yaml side.

How so? The world yaml definitions are virtually identical to what's in the PR now, except it reuses the same Country trait instead of adding a new one.

Phrohdoh commented:
Someone may make a random that contains another random.

I had another reason too but it has escaped me now.

Phrohdoh commented:
Closing per the conversation on IRC.
